From 9d75e4e29d2fbf6ae0de7f963a6aa4dd78f5133a Mon Sep 17 00:00:00 2001
From: Phil Sutter <psutter@redhat.com>
Date: Thu, 1 Oct 2015 11:44:21 +0200
Subject: [PATCH] Fix changing tunnel remote and local address to any

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=684691

commit 4e4b78324f1dbcee590c01a7ac3e5ebaf20daa27
Author: Thadeu Lima de Souza Cascardo <cascardo@redhat.com>
Date:   Thu Jun 4 09:01:18 2015 -0300

    Fix changing tunnel remote and local address to any

    If a tunnel is created with a local address, you can't change it to any.

     # ip tunnel add tunl1 mode ipip remote 10.16.42.37 local 10.16.42.214 ttl 64
     # ip tunnel show tunl1
     tunl1: ip/ip  remote 10.16.42.37  local 10.16.42.214  ttl 64
     # ip tunnel change tunl1 local any
     # echo $?
     0
     # ip tunnel show tunl1
     tunl1: ip/ip  remote 10.16.42.37  local 10.16.42.214  ttl 64

    It happens that parse_args zeroes ip_tunnel_parm, and when creating the
    tunnel, it is OK to leave it as is if the address is any. However, when
    changing the tunnel, the current parameters will be read from
    ip_tunnel_parm, and local and remote address won't be zeroes anymore, so
    it needs to be explicitly set to any.

    Signed-off-by: Thadeu Lima de Souza Cascardo <cascardo@redhat.com>
    Acked-by: Nicolas Dichtel <nicolas.dichtel@6wind.com>

Signed-off-by: Phil Sutter <psutter@redhat.com>
---
 ip/iptunnel.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/ip/iptunnel.c b/ip/iptunnel.c
index dcd7ee6..fed409f 100644
--- a/ip/iptunnel.c
+++ b/ip/iptunnel.c
@@ -159,10 +159,14 @@ static int parse_args(int argc, char **argv, int cmd, struct ip_tunnel_parm *p)
 			NEXT_ARG();
 			if (strcmp(*argv, "any"))
 				p->iph.daddr = get_addr32(*argv);
+			else
+				p->iph.daddr = htonl(INADDR_ANY);
 		} else if (strcmp(*argv, "local") == 0) {
 			NEXT_ARG();
 			if (strcmp(*argv, "any"))
 				p->iph.saddr = get_addr32(*argv);
+			else
+				p->iph.saddr = htonl(INADDR_ANY);
 		} else if (strcmp(*argv, "dev") == 0) {
 			NEXT_ARG();
 			strncpy(medium, *argv, IFNAMSIZ-1);
-- 
1.8.3.1

